proxy/enfuse/websocket_test.go

```diff
@@ -1,64 +1,95 @@
 package enfuse_test
 
 import (
-	"fmt"
 	"github.com/enfabrica/enkit/proxy/enfuse"
+	"github.com/enfabrica/enkit/proxy/enfuse/testserver"
 	"github.com/gorilla/websocket"
 	"github.com/stretchr/testify/assert"
-	"net/http"
-	"net/http/httptest"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 )
 
-func makeEchoServer(t *testing.T) *httptest.Server {
-	upg := websocket.Upgrader{}
-	m := http.NewServeMux()
-	m.HandleFunc("/echo", func(writer http.ResponseWriter, request *http.Request) {
-		webConn, err := upg.Upgrade(writer, request, nil)
-		if err != nil {
-			assert.NoError(t, err)
-			http.Error(writer, "internal test error", 500)
-			return
-		}
-		for {
-			m, data, err := webConn.ReadMessage()
-			if err != nil {
-				fmt.Println(err.Error())
-				continue
-			}
-			assert.NoError(t, webConn.WriteMessage(m, append([]byte("hello"), data...)))
-		}
-	})
-	return httptest.NewServer(m)
+func TestBasicPoolSet(t *testing.T) {
+	p := enfuse.NewPool(enfuse.DefaultPayloadStrategy)
+	assert.False(t, p.ServerPresent())
+	server1 := &websocket.Conn{}
+	assert.NoError(t, p.SetServer(server1))
+	assert.True(t, p.ServerPresent())
 }
 
-func TestSanityPool(t *testing.T) {
+func TestFetchReturnNilWithNoClients(t *testing.T) {
 	p := enfuse.NewPool(enfuse.DefaultPayloadStrategy)
-	uidLen, factoryUIDFunc := enfuse.DefaultPayloadStrategy()
+	uidLen, _ := enfuse.DefaultPayloadStrategy()
+	assert.Nil(t, p.Fetch(make([]byte, uidLen)))
+}
+
+func TestWritesToServer(t *testing.T) {
+	strategy := testserver.DoubleDigitPayloadStrategy
+	p := enfuse.NewPool(strategy)
+	uidLen, _ := strategy()
+
 	assert.Nil(t, p.Fetch(make([]byte, uidLen)))
 	assert.False(t, p.ServerPresent())
 
-	s := makeEchoServer(t)
+	var recvMessages []string
+	mu := sync.Mutex{}
+	s := testserver.NewWebsocketCounterServer(t, func(input []byte) {
+		mu.Lock()
+		recvMessages = append(recvMessages, string(input))
+		mu.Unlock()
+	})
 	defer s.Close()
 
 	url := strings.Replace(s.URL, "http", "ws", -1)
-	serverConn, _, err := websocket.DefaultDialer.Dial(url+"/echo", nil)
-	assert.NoError(t, err)
+	serverConn, _, err := websocket.DefaultDialer.Dial(url, nil)
 
+	assert.NoError(t, err)
 	assert.NoError(t, p.SetServer(serverConn))
+
 	assert.True(t, p.ServerPresent())
-	go func() {
-		for _ = range make([]int, 100){
-			go assert.NoError(t, p.SetServer(serverConn))
-			go assert.True(t, p.ServerPresent())
-		}
-	}()
-	time.Sleep(2 * time.Second)
+	messages := []string{"winnie", "piglet", "eeyore", "tigger"}
+	for _, message := range messages {
+		cc, _, err := websocket.DefaultDialer.Dial(url, nil)
+		assert.NoError(t, err)
+		err = p.WriteWebsocketServer(websocket.BinaryMessage, []byte(message), enfuse.NewWebsocketLock(cc))
+		assert.NoError(t, err)
+	}
+	// give time for cpu cycles
+	time.Sleep(200 * time.Millisecond)
```

Comment:
(nit) this scares me - most tests I see with sleeps are either flaky or slow; even tests that should work with a particular timeout experience failures when you change the build machine and/or load up the build machine (like when tests run in parallel with other builds/tests). Ideally there's some sort of condition to wait for in order to proceed with confidence.

No action needed, but hopefully the number of tests that rely on timing can be minimized moving forward

Reply:
I don't like to too, it feels way to unstable for me. I'm hoping to expand the testserver package to check and wait for all write, but that'll be another PR.